main.ts

```diff
@@ -1,150 +1,293 @@
+    this.registerEvent(this.app.vault.on('rename', this.handleRename));
+    this.registerEvent(this.app.vault.on('delete', this.handleDelete));
+    this.registerEvent(this.app.metadataCache.on('changed', this.handleMetaChange));
+    //this.registerEvent(this.app.metadataCache.on('resolve', this.handleMetaResolve));
+  }
+
+  // Restore original title on unload.
+  onunload() {
+    console.log(`unloading ${this.manifest.id} plugin`);
+    //console.log(`reverting title to '${this.baseTitle}'`);
+    document.title = this.baseTitle;
+  }
+
+  // The main method that is responsible for updating the title
+  refreshTitle(file?: TFile): void {
+    let template: any;
+    if (!file) {
+      file = this.app.workspace.getActiveFile() || undefined;
+    }
+    // For the template, the vault and workspace are always available
+    template = {
+      'vault': this.app.vault.getName(),
+      'version': (this.appVer || ''),
+      'workspace': this.app.internalPlugins.plugins.workspaces.instance.activeWorkspace // Defaults to: '' if not enabled
+    };
+    if (file instanceof TFile) {
+      // If a file is open, the filename, path and frontmatter is added
+      let cache = this.app.metadataCache.getFileCache(file);
+      if (cache && cache.frontmatter) {
+        const isTemplate = new RegExp('<%');
+        for (const [frontmatterKey, frontmatterValue] of Object.entries(cache.frontmatter || {})) {
+          let k = ('frontmatter.' + frontmatterKey) as string;
+          if (!isTemplate.test(frontmatterValue)) {
+            template[k] = frontmatterValue;
+          }
+        }
+      }
+      let friendlyBasename: string = file.basename;
+      if (file.extension !== 'md') {
+        friendlyBasename = file.name;
+      }
+      template = {
+        'filepath': file.path,
+        'filename': file.name,
+        'basename': friendlyBasename,
+        'extension': file.extension,
+        ...template
+      }
+      //console.log(template)
+      document.title = this.templateTitle(template, this.settings.titleTemplate);
+    } else {
+      document.title = this.templateTitle(template, this.settings.titleTemplateEmpty);
+    }
+  }
+
+  templateTitle(template: any, title: string): string {
+    let delimStr = this.settings.delimStr;
+    let titleSeparator = this.settings.titleSeparator;
+    // Process each template key
+    Object.keys(template).forEach(field => {
+      const hasField = new RegExp(`{{${field}}}`);
+      //console.log(`%cchecking if ${title} contains {{${field}}}`, 'background: #222; color: #a0ffff');
+      //console.log('bool: ' + hasField.test(title));
+      if (hasField.test(title) && template[field] !== null && template[field].length > 0) {
+        //console.log(`%cexecuting transforms: [${field}] --> [${template[field]}]`, 'background: #222; color: #bada55');
+        let re = new RegExp(`{{${field}}}`);
+        title = title.replace(re, `${template[field]}`);
+      }
+    });
+    // clean up delimiters
```

Comment:
I could not get this to clean up properly I think, if the template is ending with a delimiter, is it supposed to be removed right? That didn't work for me...

Reply:
no, template should not end with a delimiter... (what would the purpose be anyway?) -- see comment above

Reply:
Ah, I see where it went wrong. If you put the placeholder for the delimiter within the template with a space before it, ie:

`{{key}} ~~` the delimiter placeholder it isn't replaced. I'm not sure this is a bug or behaviour we want..

We sure need some examples in the readme, which I can add!

Reply:
Yes, that is intentional, in case the user wants to insert their own whitespace or other chars, it won't be touched. If you want a space, put it in the delimStr itself e.g. ` - ` or ` | `) 

Agree that some examples would be helpful- if you don't have time I can also help to create that